torch/csrc/jit/api/function_impl.cpp

```diff
@@ -72,12 +73,25 @@ const c10::FunctionSchema& GraphFunction::getSchema() const {
 
 void preoptimizeGraph(std::shared_ptr<Graph>& graph) {
   Inline(*graph);
+
   // Peephole Optimize cleans up many "is None" checks and creates constant prop
   // opportunities
   PeepholeOptimize(graph, true);
-  // // AliasDb construction can be slow, so run it just on immutable types
-  // // to clean up constant Ifs & other easy wins
+
+  // AliasDb construction can be slow, so run it just on immutable types
+  // to clean up constant Ifs & other easy wins
   ConstantPropagationImmutableTypes(graph);
+
+  // Inject casts for automatic mixed precision
+  //
+  // TODO: Ideally, this pass could run earlier, before inlining
+  //  or any other optimizations. That setup is preferable because:
+  //  1. The AMP pass would be self-contained and function independently 
+  //     of the any optimizations
+  //  2. AMP transformations would benefit from followup passes's cleanup
+  //
+  Autocast(graph);
```

Comment:
I'm a little suspicious on our choice to insert the pass. `preoptimizeGraph` is called when we inline `CallFunction` as well as `CallMethod`.

Wouldn't it result in nested `autocast` to take effect multiple times?

Reply:
Good point. Functionally this is not a problem, as I mentioned in a previous comment. 

The current placement of the pass is convenient for the proof of concept, but I agree, it has a number of drawbacks. This is why I'm not rushing to productize it just yet.